Restructure Api Gateway Documentation (#13675)
hashicorp/consul: website/content/docs/api-gateway/usage/basic-usage.mdx

---
layout: docs
page_title: Consul API Gateway Basic Usage
description:  >-
  Consul API Gateway Basic Usage
---

# Basic Usage

1. Verify that the [requirements](/docs/api-gateway/consul-api-gateway-install#requirments) have been met.
1. Verify that the Consul API Gateway CRDs and controller have been installed and applied (see [Installation](/docs/api-gateway/consul-api-gateway-install)).

> 1. Configure the artifacts described below in [Configuration](/docs/api-gateway/configuration).

Reviewer: ```suggestion
1. Configure the artifacts as describe in the [Configuration](/docs/api-gateway/configuration) section.
```

<CodeBlockConfig hideClipboard filename="values.yaml">

  ```yaml

>   apiGateway:
>   managedGatewayClass:
>   enabled: true

Reviewer 2: Indentation is a little off here, but I'm not sure if we should keep this section. What is it intended to demonstrate? I tripped on the empty `managedGatewayClass`

```suggestion
  apiGateway:
    enabled: true
    managedGatewayClass:
```

  ```

</CodeBlockConfig>

1. Issue the `kubectl apply` command to implement the configurations, e.g.:

```shell-session

> $ kubectl apply -f gateway.yaml routes.yaml

Reviewer 2: ```suggestion
$ kubectl apply --filename gateway.yaml routes.yaml
```

>    ```

Reviewer 2: Indentation here is inconsistent, but I'm not 100% sure which level is what we want

| `-pprof`               | Integer value that specifies the Go pprof port number for collecting metrics.                                                                                                         | Optional | none                                                                    |
| `-sds-server-host`     | String value that specifies the host server for the secret discovery service (SDS).                                                                                                   | Optional | `consul-api-gateway-controller.default.`<br/>`svc.cluster.`<br/>`local` |
| `-sds-server-host`     | Integer value that specifies the port number for the secret discovery service (SDS).                                                                                                  | Optional | `9090`                                                                  |

You can also issue the `version` command to print the Consul API Gateway version to the console:

```shell-session
$ ./consul-api-gateway version
consul-api-gateway 0.1.0
```
--->

> ## Common Error Messages
> 
> Some of the errors messages commonly encountered during installation and operations of Consul API Gateway are listed below, along with suggested methods for resolving them.
> 

Reviewer: ```suggestion
## Error Messages

```
A few things - 

First, I think we should just talk about "error messages" as opposed to "common error messages" because I don't want to create the impression that people are always running into a lot of errors. It's a bit rhetorical, but if the idea is to create a catalog of error messages, describing them as _common_, IMO, implies that errors are common.  

Second, I suggest trimming the first paragraph because it doesn't seem to add a lot of value. 

Third, I think I was wrong about suggesting that this section be merged into this page. I think it's fine for now, but when we build the catalog of error messages it might make sense for them to exist as a standalone topic. I'm open to discussion.

the text of the error.

Two examples:
    ### Failed opening file during installation
    ### Message: "Can't connect to repository" when running Helm chart

******** Template for new Error Messages ********
 Copy and paste the following 13 lines when adding a new error message to this page.

### Title for this error

```
Replace with text of example error message
```
**Conditions:**
REPLACE THIS with description of when and why the error is typically seen

**Impact:**
REPLACE THIS with description of most likely impact of the event that caused this error to occur

**Recommended Action:**
REPLACE THIS with the actions the user should take to try to correct the situation

***************************************************************************
--->

### Helm installation failed: "no matches for kind"

```log
Error: INSTALLATION FAILED: unable to build kubernetes objects from release manifest: [unable to recognize "": no matches for kind "GatewayClass" in version "gateway.networking.k8s.io/v1alpha2", unable to recognize "": no matches for kind "GatewayClassConfig" in version "api-gateway.consul.hashicorp.com/v1alpha1"]
```
**Conditions:**

> When this error occurs during the process of installing Consul API Gateway, it is usually caused by not having the required CRD files installed in Kubernetes prior to installing Consul API Gateway.

Reviewer: ```suggestion
Consul API Gateway generates this error when the required CRD files have not been installed in Kubernetes prior to installing Consul API Gateway. 
```
Trying to restate in active voice. Is this still accurate? Does this message only appear during the installation step?

Author: I believe it should only appear during the installation step, but @nathancoleman @andrewstucki @mikemorris can one of you confirm?

**Impact:**

> The installation process will typically fail after this error message is generated

Reviewer: ```suggestion
The installation process typically fails after this error message is generated.
```

---
layout: docs
page_title: Consul API Gateway Basic Usage
description:  >-
  Consul API Gateway Basic Usage
---

> 

Reviewer 3: Since this is the only item in "Usage", can we flatten it and just call this one "usage"?

<img width="272" alt="image" src="https://user-images.githubusercontent.com/7556325/182292114-e18346c0-ce0d-4660-aac3-6f67ec5bd12d.png">

Author: d23cff615f5bd3a0ae99ef3895dcbb41cba0c5e8

# Basic Usage

This topic describes the basic workflow for implementing Consul API Gateway configurations.

> 1. Verify that the [requirements](/docs/api-gateway/consul-api-gateway-install#requirments) have been met.

Reviewer 3: i believe we deprecated this page. also, can we list the requirements here so users aren't clicking on a different page? this might reduce conversion rates

```suggestion
1. Verify that the [requirements](/docs/api-gateway/tech-specs) have been met.
```

Author: I had originally moved the requirements onto this page, but had been asked to move them back to the separate page.

1. Verify that the Consul API Gateway CRDs and controller have been installed and applied (see [Installation](/docs/api-gateway/consul-api-gateway-install)).
1. Configure the artifacts as describe in the [Configuration](/docs/api-gateway/configuration) section.

<CodeBlockConfig hideClipboard filename="values.yaml">

  ```yaml
  apiGateway:
    enabled: true
    managedGatewayClass:
  ```

> </CodeBlockConfig>

Reviewer 3: is this the wrong place? this doesn't feel right. also, can we provide a high level working example? the next command references `gateway.yaml` and `routes.yaml`, neither of which was defined

ideally, we can replace this page with a "Get Started" that links out to the API Gateway learn tutorial (https://learn.hashicorp.com/tutorials/consul/kubernetes-api-gateway). what do you think @trujillo-adam?
